File: SimpleFocuser/SimpleFocuser/SetupDialogForm.vb
Imports System.Windows.Forms
Imports System.Runtime.InteropServices
Imports ASCOM.Utilities
Imports ASCOM.TeenAstro

<ComVisible(False)>
Public Class SetupDialogForm
  Private Sub saveState()
    If RadioButtonSP.Checked Then
      Focuser.mcomPort = ComboBoxComPort.SelectedItem
      Focuser.mInterface = "COM"
    ElseIf RadioButtonIP.Checked Then
      Focuser.mIP = TextBoxIP.Text
      Focuser.mPort = TextBoxPort.Text
      Focuser.mInterface = "IP"

      If Not System.Net.IPAddress.TryParse(Focuser.mIP, Focuser.mobjectIP) Then
        MsgBox(Focuser.mIP + " is not a valid IP Address")
        Return
      End If
    End If
    Focuser.mtraceState = chkTrace.Checked
    Focuser.mStepSize = NumericUpDownStepSize.Value

  End Sub

  Private Sub OK_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OK_Button.Click ' OK button event handler
    saveState()
    Me.DialogResult = System.Windows.Forms.DialogResult.OK

    Me.Close()
  End Sub

  Private Sub Cancel_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel_Button.Click 'Cancel button event handler
    Me.DialogResult = System.Windows.Forms.DialogResult.Cancel
    Me.Close()
  End Sub

  Private Sub SetupDialogForm_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load ' Form load event handler
    ' Retrieve current values of user settings from the ASCOM Profile
    InitUI()
  End Sub

  Private Sub InitUI()
    Dim assembly As System.Reflection.Assembly = System.Reflection.Assembly.GetExecutingAssembly()
    Dim fvi As FileVersionInfo = FileVersionInfo.GetVersionInfo(assembly.Location)
    Dim version As String = fvi.FileVersion
    Me.Text = "TeenAstro Focuser " + version
    chkTrace.Checked = Focuser.mtraceState
    TextBoxIP.Text = Focuser.mIP
    TextBoxPort.Text = Focuser.mPort
    RadioButtonSP.Checked = Focuser.mInterface = "COM"
    RadioButtonIP.Checked = Focuser.mInterface = "IP"
    ' set the list of com ports to those that are currently available
    ComboBoxComPort.Items.Clear()
    ComboBoxComPort.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames())       ' use System.IO because it's static
    ' select the current port if possible
    If ComboBoxComPort.Items.Contains(Focuser.mcomPort) Then
      ComboBoxComPort.SelectedItem = Focuser.mcomPort
    End If
    NumericUpDownStepSize.Value = Focuser.mStepSize
  End Sub

  Private Sub ButtonEEPROM_Click(sender As Object, e As EventArgs) Handles ButtonEEPROM.Click
    saveState()
    Focuser.WriteProfile()
    Dim FormEEPROM As ConfigEEPROM = New ConfigEEPROM()
    FormEEPROM.ShowDialog()
  End Sub

  Private Sub PictureBoxASCOM_Click(sender As Object, e As EventArgs) Handles PictureBoxASCOM.Click
    Try
      System.Diagnostics.Process.Start("http://ascom-standards.org/")
    Catch noBrowser As System.ComponentModel.Win32Exception
      If noBrowser.ErrorCode = -2147467259 Then
        MessageBox.Show(noBrowser.Message)
      End If
    Catch other As System.Exception
      MessageBox.Show(other.Message)
    End Try
  End Sub

  Private Sub PictureBoxTeenAstro_Click(sender As Object, e As EventArgs) Handles PictureBoxTeenAstro.Click
    Try
      System.Diagnostics.Process.Start("https://groups.io/g/TeenAstro")
    Catch noBrowser As System.ComponentModel.Win32Exception
      If noBrowser.ErrorCode = -2147467259 Then
        MessageBox.Show(noBrowser.Message)
      End If
    Catch other As System.Exception
      MessageBox.Show(other.Message)
    End Try
  End Sub

  Private Sub RadioButtonIP_CheckedChanged(sender As Object, e As EventArgs) Handles RadioButtonIP.CheckedChanged
    TextBoxIP.Enabled = RadioButtonIP.Checked
    TextBoxPort.Enabled = RadioButtonIP.Checked
  End Sub

  Private Sub RadioButtonSP_CheckedChanged(sender As Object, e As EventArgs) Handles RadioButtonSP.CheckedChanged
    ComboBoxComPort.Enabled = RadioButtonSP.Checked
  End Sub

  Private Sub TextBoxPort_KeyPress(sender As Object, e As KeyPressEventArgs) Handles TextBoxPort.KeyPress
    If Not Char.IsDigit(e.KeyChar) And Not Char.IsControl(e.KeyChar) Then
      e.Handled = True
    End If
  End Sub


End Class

File: TeenAstroASCOM_V71/TeenAstroASCOM_V71/TelescopeDriver/SetupDialogForm.cs
using ASCOM.Utilities;
using System;
using System.ComponentModel;
using System.Runtime.InteropServices;
using System.Windows.Forms;

namespace ASCOM.TeenAstro.Telescope
{
    [ComVisible(false)] // Form not registered for COM!
    public partial class SetupDialogForm : Form
    {
        const string NO_PORTS_MESSAGE = "No COM ports found";
        TraceLogger tl; // Holder for a reference to the driver's trace logger

        public SetupDialogForm(TraceLogger tlDriver)
        {
            InitializeComponent();

            // Save the provided trace logger for use within the setup dialogue
            tl = tlDriver;

            // Initialise current values of user settings from the ASCOM Profile
            InitUI();
        }

        private void CmdOK_Click(object sender, EventArgs e) // OK button event handler
        {
            // Place any validation constraint checks here and update the state variables with results from the dialogue

            tl.Enabled = chkTrace.Checked;

            if (RadioButtonSP.Checked)
            {
                // Update the COM port variable if one has been selected
                if (comboBoxComPort.SelectedItem is null) // No COM port selected
                {
                    tl.LogMessage("Setup OK", $"New configuration values - COM Port: Not selected");
                }
                else if (comboBoxComPort.SelectedItem.ToString() == NO_PORTS_MESSAGE)
                {
                    tl.LogMessage("Setup OK", $"New configuration values - NO COM ports detected on this PC.");
                }
                else // A valid COM port has been selected
                {
                    tl.LogMessage("Setup OK", $"New configuration values - COM Port: {comboBoxComPort.SelectedItem}");
                    TelescopeHardware.comPort = comboBoxComPort.SelectedItem.ToString();
                    TelescopeHardware.Interface = TelescopeHardware.InterfaceCOM;
                }
            }
            else if (RadioButtonIP.Checked)
            {
                TelescopeHardware.IP = TextBoxIP.Text;
                TelescopeHardware.Port = Convert.ToInt16(TextBoxPort.Text);
                TelescopeHardware.Interface = TelescopeHardware.InterfaceIP;
            }

            this.DialogResult = DialogResult.OK;
            this.Close();
        }

        private void CmdCancel_Click(object sender, EventArgs e) // Cancel button event handler
        {
            Close();
        }

        private void BrowseToAscom(object sender, EventArgs e) // Click on ASCOM logo event handler
        {
            try
            {
                System.Diagnostics.Process.Start("https://ascom-standards.org/");
            }
            catch (Win32Exception noBrowser)
            {
                if (noBrowser.ErrorCode == -2147467259)
                    MessageBox.Show(noBrowser.Message);
            }
            catch (Exception other)
            {
                MessageBox.Show(other.Message);
            }
        }

        private void InitUI()
        {
            chkTrace.Checked = tl.Enabled;
            TextBoxIP.Text = TelescopeHardware.IP;
            TextBoxPort.Text = TelescopeHardware.Port.ToString();
            RadioButtonSP.Checked = TelescopeHardware.Interface == TelescopeHardware.InterfaceCOM;
            RadioButtonIP.Checked = TelescopeHardware.Interface == TelescopeHardware.InterfaceIP;

            comboBoxComPort.Items.Clear();
            using (Serial serial = new Serial())
            {
                comboBoxComPort.Items.AddRange(serial.AvailableCOMPorts);
            }

            if (comboBoxComPort.Items.Count == 0)
            {
                comboBoxComPort.Items.Add(NO_PORTS_MESSAGE);
                comboBoxComPort.SelectedItem = NO_PORTS_MESSAGE;
            }

            if (comboBoxComPort.Items.Contains(TelescopeHardware.comPort))
                comboBoxComPort.SelectedItem = TelescopeHardware.comPort;

            tl.LogMessage("InitUI", $"Trace: {chkTrace.Checked}, COM: {comboBoxComPort.SelectedItem}, IP: {TextBoxIP.Text}");
        }

        private void SetupDialogForm_Load(object sender, EventArgs e)
        {
            if (WindowState == FormWindowState.Minimized)
                WindowState = FormWindowState.Normal;
            else
            {
                TopMost = true;
                Focus();
                BringToFront();
                TopMost = false;
            }
        }

        private void RadioButtonIP_CheckedChanged(object sender, EventArgs e)
        {
            TextBoxIP.Enabled = RadioButtonIP.Checked;
            TextBoxPort.Enabled = RadioButtonIP.Checked;
        }

        private void RadioButtonSP_CheckedChanged(object sender, EventArgs e)
        {
            comboBoxComPort.Enabled = RadioButtonSP.Checked;
        }

        private void TextBoxPort_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (!char.IsDigit(e.KeyChar) && !char.IsControl(e.KeyChar))
                e.Handled = true;
        }
    }
}